lib/scheduler.js

```diff
@@ -258,8 +258,18 @@ module.exports = class Scheduler {
     const jobName  = job.jobName;
     const time     = job.time;
     const schedule = this._schedules.get(jobName);
-    if (schedule)
-      return schedule._runJob(jobName, time);
+    if (schedule) {
+      // Skip recurring jobs outside of their run window, most
+      // commonly because the job failed and was returned
+      // to the queue.
+      const jobExpired = jobIsExpired(schedule, job);
```

Comment:
This function is responsible for running the scheduled job.

Sometimes. There's some logic that determines whether the job should run, and this function acts based on that logic. So it would look like:

`if (shouldRunJob(schedule, job)) {`

(That also results in an if statement, where the condition is positive, and the most common case comes first)

`shouldRunJob` decides whether the job should run or not. To `_runQueuedJob`, the logic that makes this decision is an implementation detail; it only cares about the outcome: run or not.

The decision can be that jobs only run between the hours of 10pm and 8am, or only limit to three attempts (if we wanted to limit retries), or only run when they don't overlap. So this function can make a lot of different decisions that to `_runQueuedJob` are all the same.

At the moment, we're only looking to avoid running overlapping scheduled jobs. But even that can be implemented in so many ways: we can use locks, counters, or decide each job has a time to live.

Having `shouldRunJob` has an abstraction allows us to change what decision we make, or how we make that decision, without having to touch `_runQueuedJob`. It doesn't leak.

Side bonus, we don't have to worry about `jobExpired`, `jobIsExpired` and `isExpired` all holding the same value, but having to innovate on name, so they don't shadow each other.